### ISSUE TITLE
lazy module loading is broken in multithreaded scenario

### ISSUE THREAD

**Issue by eindenbom** [librosa/librosa#1799]
When lazy loaded module (e.g. resampy) is concurrently accessed from multiple threads, halfway loaded module is used leading errors like:

```
AttributeError: module 'resampy' has no attribute 'resample'module 'resampy' has no attribute 'resample'
```

I suggest disabling lazy loading until fixed in importlib.

**To Reproduce**
Run the following snippet:
```
import librosa
import threading
import numpy as np


def resample():
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")


for _ in range(5):
    threading.Thread(target=resample).start()
```

**Expected behavior**
No errors reported.

**Actual output**
```
Traceback (most recent call last):
Traceback (most recent call last):
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
Traceback (most recent call last):
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
Traceback (most recent call last):
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
          File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
self._target(*self._args, **self._kwargs)self._target(*self._args, **self._kwargs)

  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
    self._target(*self._args, **self._kwargs)
  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
AttributeError: module 'resampy' has no attribute 'resample'
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
AttributeError: module 'resampy' has no attribute 'resample'AttributeError
: module 'resampy' has no attribute 'resample'
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
AttributeError: module 'resampy' has no attribute 'resample'
python-BaseException
python-BaseException
python-BaseException
python-BaseException
Exception in thread Exception in thread Thread-8:
Exception in thread Traceback (most recent call last):
Thread-5:
Traceback (most recent call last):
  File "C:\Build\envs\dec5\lib\threading.py", line 932, in _bootstrap_inner
  File "C:\Build\envs\dec5\lib\threading.py", line 932, in _bootstrap_inner
Thread-9:
Traceback (most recent call last):
  File "C:\Build\envs\dec5\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
    self.run()
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
        self._target(*self._args, **self._kwargs)
self.run()
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
    self._target(*self._args, **self._kwargs)
  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    self._target(*self._args, **self._kwargs)
  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
AttributeError: AttributeError: module 'resampy' has no attribute 'resample'
module 'resampy' has no attribute 'resample'    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")

  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
python-BaseException
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
AttributeError: module 'resampy' has no attribute 'resample'
Exception in thread Thread-6:
Traceback (most recent call last):
  File "C:\Build\envs\dec5\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Build\envs\dec5\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "D:\source\internal_ai_deepaudio\training\break_lazy_load.py", line 7, in resample
    librosa.resample(np.zeros(1000), orig_sr=24_000, target_sr=16_000, res_type="kaiser_best")
  File "C:\Build\envs\dec5\lib\site-packages\librosa\core\audio.py", line 677, in resample
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=axis)
AttributeError: module 'resampy' has no attribute 'resample'
```

**Software versions**
```
Windows-10-10.0.22621-SP0
Python 3.8.0 (default, Nov  6 2019, 16:00:02) [MSC v.1916 64 bit (AMD64)]
NumPy 1.23.5
SciPy 1.10.1
librosa 0.10.1
INSTALLED VERSIONS
------------------
python: 3.8.0 (default, Nov  6 2019, 16:00:02) [MSC v.1916 64 bit (AMD64)]

librosa: 0.10.1

audioread: 3.0.1
numpy: 1.23.5
scipy: 1.10.1
sklearn: 1.3.2
joblib: 1.3.2
decorator: 5.1.1
numba: 0.58.1
soundfile: 0.12.1
pooch: v1.8.0
soxr: 0.3.7
typing_extensions: installed, no version number available
lazy_loader: installed, no version number available
msgpack: 1.0.7

numpydoc: None
sphinx: None
sphinx_rtd_theme: None
matplotlib: None
sphinx_multiversion: None
sphinx_gallery: None
mir_eval: None
ipython: None
sphinxcontrib.rsvgconverter: None
pytest: None
pytest_mpl: None
pytest_cov: None
samplerate: None
resampy: 0.4.2
presets: None
packaging: 23.2
```

**Comment by bmcfee** [comment]
I think you should report this upstream to the lazy loader maintainers, as it's out of scope for us to fix here.

As a workaround though, have you tried calling the necessary functions first before launching threads?

**Comment by eindenbom** [comment]
I've reported an issue here: https://github.com/scientific-python/lazy_loader/issues/88

I still do not think that it is right to use in librosa functionality that is known to be broken. Lazy loading should be disabled until lazy_loaded is fixed.

PS As a workaround I preload resampy, but I'd prefer to have a real fix as it's nowhere exposed which modules could be used by librosa.

**Comment by bmcfee** [comment]
> I still do not think that it is right to use in librosa functionality that is known to be broken. Lazy loading should be disabled until lazy_loader is fixed.

I disagree.  It's impractical for us to put out patch releases every time an upstream dependency has a minor bug.  Moreover, since lazy_loader is also used by one of our other upstream dependencies (scipy), patching it out of librosa wouldn't solve your problem anyway.


> As a workaround I preload resampy, but I'd prefer to have a real fix as it's nowhere exposed what modules could be used by librosa.

A better alternative, if you don't care at all about lazy-loader, is to force eager imports as documented here: https://github.com/scientific-python/lazy_loader#early-failure  - this has the advantage of not requiring any code modification, so you can simply revert the environment variable setting once the issue is resolved.

**Comment by eindenbom** [comment]
I disagree on the classification of this problem as a minor bug in dependency. Multithreaded processing is very important and widely used, so it should have been tested in both lazy_loader and librosa.

It's up to maintainers to decide whether to release a patch release to address a particular problem, but MT tests should be added anyway when lazy_loader fix will be available or lazy loading will become part of python itself.

**Comment by bmcfee** [comment]
> I disagree on the classification of this problem as a minor bug in dependency.

It's minor in the sense that it can be worked around (from our side) without code modifications.


> Multithreaded processing is very important and widely used, so it should have been tested in both lazy_loader and librosa.


I don't want to start a huge argument over this, but is `threading` actually widely used in scientific python contexts?  My experience here has always favored multi-processing (via multiprocessing, joblib, dask, etc) over threading due to the separated state and GIL issues.  I'm not usually looking to do compute while IO is blocking (the main benefit of threading within one python runtime), but rather just get multiple things done in parallel (which `threading` doesn't actually do).  I'm sure you have valid reasons for going with threading here, but my hunch is that for our typical use cases, multiprocessing pays off more than threading.  (See: https://effectivepython.com/ chapter 53, "Use Threads for Blocking I/O, Avoid for Parallelism" for more discussion on this point.)

**Comment by bmcfee** [comment]
I'm closing this out as an upstream issue.  Beyond suggesting workarounds, I don't think there's anything to be done here from the librosa side.
